### PR TITLE
Netcore satellite assemblies

### PR DESCRIPTION
**Pull request by cdmihai**
- reuse the full framework satellites in netcore / enable localized netcore builds (only on windows, after a full framework build)
- add a new nuget package, `Microsoft.Build.Localization` which contains the satellites
- redo the build output to be consistent between full framework and netcore

Need to
- [x] update vsix generation to new layout
- [x] make CI do localized netcore builds on Windows
- [x] update our microbuild definition to new layout and to produce localized netcore builds
- [x] do more testing on non-Windows machines to make sure `dotnet build` copies everything
- [x] do more testing on non-English machines

### CODE REVIEW

**Comment by cdmihai** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please

**Comment by cdmihai** [comment]
@dotnet-bot test OSX Build for CoreCLR please

No way these changes cause the following failure https://ci2.dot.net/job/Microsoft_msbuild/job/master/job/innerloop_OSX_CoreCLR_prtest/74/
(famous last words)

**Comment by cdmihai** [comment]
@Microsoft/msbuild-maintainers 
Is everyone fine if I don't include the satellites for Conversion, legacy engine, and MSBuildTaskHost inside the localization package?

**Comment by cdmihai** [comment]
@dotnet-bot test ci please

**Comment by cdmihai** [comment]
@dotnet-bot test this please

**Comment by cdmihai** [comment]
@dotnet-bot help

**Comment by cdmihai** [comment]
@dotnet-bot test ci please

**Comment by cdmihai** [comment]
@dotnet-bot test OSX Build for CoreCLR please

@rainersigwald have you seen these types of osx failures before? https://ci2.dot.net/job/Microsoft_msbuild/job/master/job/innerloop_OSX_CoreCLR_prtest/79/

**Comment by cdmihai** [comment]
@dotnet-bot test OSX Build for CoreCLR please

**Comment by cdmihai** [comment]
@dotnet-bot test OSX Build for CoreCLR please

**Comment by cdmihai** [comment]
@dotnet-bot test OSX Build for CoreCLR please

**Comment by cdmihai** [comment]
@dotnet-bot test OSX Build for CoreCLR please

**Comment by cdmihai** [comment]
@dotnet-bot test OSX Build for CoreCLR please

**Comment by cdmihai** [comment]
@Microsoft/msbuild-maintainers This is ready for review

**Comment by cdmihai** [comment]
@dotnet-bot test OSX Build for CoreCLR please
@dotnet-bot test Windows_NT Build for Desktop please

File: netci.groovy [hunk 1]
```diff
@@ -32,14 +32,20 @@ def branch = GithubBranchName
...
 
-                            // only Desktop support localized builds 
-                            if (runtime == "Desktop") {
-                                windowsScript += " --localized-build"
+                            if (runtime == "Full") {
+                                script += " --scope Test"
+                            }
+                            // .net core builds are localized (they need the satellites from the full framework build), run tests, and also build the nuget packages
+                            else if (runtime == "CoreCLR") {
+                                script == " --scope Compile --build-only"
```
Comment (rainersigwald) [inline netci.groovy]: Explicitly add `--target Full` here?

File: netci.groovy [hunk 2]
```diff
@@ -32,14 +32,20 @@ def branch = GithubBranchName
...
+                                script += " && cibuild.cmd --target CoreCLR --scope test --localized-Build"
+                                script += " && msbuild build/NugetPackages/CreateNuGetPackages.proj"
```
Comment (rainersigwald) [inline netci.groovy]: I'd be more comfortable if this was all baked into `cibuild.cmd` so it'd be easier to repro on a local machine.

File: src/Build/AssemblyInfo.cs [hunk 3]
```diff
@@ -43,10 +43,10 @@
 // Needed for the "hub-and-spoke model to locate and retrieve localized resources": https://msdn.microsoft.com/en-us/library/21a15yht(v=vs.110).aspx
 // We want "en" to require a satellite assembly for debug builds in order to flush out localization
 // issues, but we want release builds to work without it.
-#if (DEBUG)
+#if (DEBUG && !RUNTIME_TYPE_NETCORE)
```
Comment (rainersigwald) [inline src/Build/AssemblyInfo.cs]: Can  you update the comment with reasoning?

File: src/dir.targets [hunk 4]
```diff
@@ -108,4 +110,55 @@
       </ReferencePath>
     </ItemGroup>
   </Target>
+
+  <!-- copied from Microsoft.Common.CurrentVersion.targets and modified to force anycpy -->
```
Comment (rainersigwald) [inline src/dir.targets]: typo: cpy

File: targets/runtimeDependencies/project.json [hunk 5]
```diff
@@ -46,6 +46,7 @@
         "System.Runtime.Extensions": "4.1.0",
         "System.Runtime.InteropServices.RuntimeInformation": "4.0.0",
         "System.Runtime.Loader": "4.0.0",
+	"System.Text.Encoding.CodePages" : "4.0.1",
```
Comment (rainersigwald) [inline targets/runtimeDependencies/project.json]: nit: alignment

File: src/MSBuild/XMake.cs [hunk 6]
```diff
@@ -1334,8 +1334,15 @@ internal static void SetConsoleUI()
...
             {
                 thisThread.CurrentUICulture = new CultureInfo("en-US");
+                return;
             }
 #endif
+#if RUNTIME_TYPE_NETCORE
+            // https://github.com/dotnet/roslyn/issues/10785#issuecomment-238940601
+            // by default, .NET Core doesn't have all code pages needed for Console apps.
+            // see the .NET Core Notes in https://msdn.microsoft.com/en-us/library/system.diagnostics.process(v=vs.110).aspx
+            Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
```
Comment (nguerrera) [inline src/MSBuild/XMake.cs]: This shouldn't be necessary on .NET Core 2.0

File: netci.groovy (hunk 2 above)
Comment (cdmihai) [inline netci.groovy]: The baked in version would be `cibuild.cmd --target All --scope Test --localized-build && msbuild build/NugetPackages/CreateNuGetPackages.proj`
The problem there is that `--target All --scope Test` runs tests for both frameworks. I could change the script to this and then on windows jobs do not disambiguate between Full and Core as there's a command to rule them all.

Alternative is to have `cibuild.cmd --WindowsCoreLocalizedJob` which rewrites to `cibuild.cmd --target Full --scope Compile --build-only --localized-build && cibuild.cmd --target CoreCLR --scope test --localized-Build && msbuild build/NugetPackages/CreateNuGetPackages.proj`

The first option is less code, but when something fails we don't know whether it's FF or Core. The second option adds more complexity to the script. Second option then? :)

Comment (rainersigwald) [inline netci.groovy]: Second option sounds fairly good.

**Comment by nguerrera** [comment]
I tried a private drop of the packages and it seems that having the files in contentFiles is preventing them from propagating correctly to the deps.json. Pulling this down to try a different scheme locally.

**Comment by cdmihai** [comment]
@nguerrera I also tested this with:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>library</OutputType>
    <TargetFrameworks>netstandard1.3;netstandard1.5</TargetFrameworks>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Build.Runtime" Version="15.3.0-preview-000259-00"/>
    <PackageReference Include="Microsoft.Build.Localization" Version="15.3.0-preview-000259-00"/>
  </ItemGroup>

</Project>

```

Both dotnet build and publish seem to copy the satellites to the output with the correct file layout. How is your attempt different?

**Comment by cdmihai** [comment]
@dotnet-bot test ci please

**Comment by nguerrera** [comment]
The files are getting copied, but they're not getting loaded. They're missing from the MSBuild.deps.json we generate.

**Comment by cdmihai** [comment]
Is that a bug in the deps.json generation?
I see that MSBuild.exe and MSBuild.dll from `Microsoft.Build.Runtime` are also content files. Are those getting in the deps.json?

**Comment by nguerrera** [comment]
I don't think it's a bug. contentFiles are not things that are supposed to be loaded. I think MSBuild.dll works because it's the root program not a dependency of it.

**Comment by nguerrera** [comment]
I think lib\any\locale may work. I tried to test but I'm having too much difficulty building the repo right now.

Look at FluentValidation.resources in deps.json after dotnet publish of project below. We need the same for the MSBuild resource dlls:

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Build.Runtime" Version="15.3.0-preview-000255-00"/>
    <PackageReference Include="Microsoft.Build.Localization" Version="15.3.0-preview-000255-00"/>
    <PackageReference Include="FluentValidation" Version="5.5.0" />
  </ItemGroup>
</Project>
```

**Comment by cdmihai** [comment]
I tried a new package layout (`Microsoft.Build.Localization2`) where I duplicate the assemblies for all the frameworks that we're using: [Packages.zip](https://github.com/Microsoft/msbuild/files/1027484/Packages.zip)

dotnet publish seesm to create a deps.json that now contains everything. Going to look into your solution as well.

**Comment by cdmihai** [comment]
Putting them under `lib/any/<locale>` doesn't work. Nuget does not seem to understand `any` when it's under `lib`:

```
Package Microsoft.Build.Localization2 15.3.0-preview-000261-00 is not compatible with netstandard1.3 (.NETStandard,Version=v1.3). Package Microsoft.Build.Localization2 15.3.0-preview-000261-00 supports: dotnet (.NETPlatform,Version=v5.0)
      One or more packages are incompatible with .NETStandard,Version=v1.3.
      Package Microsoft.Build.Localization2 15.3.0-preview-000261-00 is not compatible with netstandard1.5 (.NETStandard,Version=v1.5). Package Microsoft.Build.Localization2 15.3.0-preview-000261-00 supports: dotnet (.NETPlatform,Version=v5.0)
      One or more packages are incompatible with .NETStandard,Version=v1.5.
      Package Microsoft.Build.Localization2 15.3.0-preview-000261-00 is not compatible with netstandard1.3 (.NETStandard,Version=v1.3) / win7-x86. Package Microsoft.Build.Localization2 15.3.0-preview-000261-00 supports: dotnet (.NETPlatform,Version=v5.0)
      One or more packages are incompatible with .NETStandard,Version=v1.3 (win7-x86).
      Package Microsoft.Build.Localization2 15.3.0-preview-000261-00 is not compatible with netstandard1.5 (.NETStandard,Version=v1.5) / win7-x86. Package Microsoft.Build.Localization2 15.3.0-preview-000261-00 supports: dotnet (.NETPlatform,Version=v5.0)
      One or more packages are incompatible with .NETStandard,Version=v1.5 (win7-x86).
```

Here are the any packages: [Packages.zip](https://github.com/Microsoft/msbuild/files/1027500/Packages.zip)

Looks like duplicating the satellites across 4 TFMs is the way to go :(

**Comment by cdmihai** [comment]
The `Nuget Package Explorer` also says `Unrecognized Framework` for `lib/any`

**Comment by nguerrera** [comment]
What if you put only netstandard1.3? All the others are compatible with that and should get assets from it fine.

**Comment by cdmihai** [comment]
That seems to work fine. Here are the packages: [Packages.zip](https://github.com/Microsoft/msbuild/files/1027654/Packages.zip)
If they work out for you, I'll merge in this PR and build new myget packages.

**Comment by nguerrera** [comment]
It's alive!

```
C:\src\cli\artifacts\win10-x64\stage2>dotnet msbuild /?
Microsoft (R) Build Engine, version 15.3.0.0
Copyright (C) Microsoft Corporation. Tous droits réservés.

Syntaxe :              MSBuild.exe [options] [fichier projet | répertoire]

Description :         Génère les cibles spécifiées dans le fichier projet. Si
                     aucun fichier projet n'est spécifié, MSBuild recherche dans le
                     répertoire de travail actuel un fichier dont
                     l'extension se termine par « proj » et utilise ce dernier. Si
                     un répertoire est spécifié, MSBuild recherche dans ce
                     répertoire un fichier projet.
```

File: build/NuGetPackages/Microsoft.Build.Localization.nuspec [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    <licenseUrl>$licenseUrl$</licenseUrl>
+    <projectUrl>$projectUrl$</projectUrl>
+    <iconUrl>$iconUrl$</iconUrl>
+    <description>This package contains the localized satellite assemblies for MSBuild. Reference this package only if you want msbuild output to be localized in cs;de;en;es;fr;it;ja;ko;pl;pt-BR;ru;tr;zh-Hans;zh-Hant.</description>
+    <copyright>Copyright © Microsoft Corporation</copyright>
+    <tags>MSBuild</tags>
+
+    <contentFiles>
+      <files include="**" buildAction="None" copyToOutput="true" flatten="false" />
+    </contentFiles>
```
Comment (nguerrera) [inline build/NuGetPackages/Microsoft.Build.Localization.nuspec]: I don't think this is needed anymore

File: build/NuGetPackages/Microsoft.Build.Localization.nuspec [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  </metadata>
+  <files>
+
+    <!-- MSBuild.dll -->
+    <file src="$outputPath$cs\MSBuild.resources.dll" target="lib\netstandard1.3\cs\MSBuild.resources.dll" buildAction="None" copyToOutput="true"/>
```
Comment (nguerrera) [inline build/NuGetPackages/Microsoft.Build.Localization.nuspec]: You probably don't need buildAction and copyToOutput.

**Comment by cdmihai** [comment]
@dotnet-bot test Windows_NT Build for Desktop please

**Comment by cdmihai** [comment]
@dotnet-bot test ci please

**Comment by nguerrera** [comment]
Is this ready to pickup via myget? Which package version should I use?

**Comment by cdmihai** [comment]
Not yet, we're waiting for some other PRs that should go in today.

**Comment by cdmihai** [comment]
@nguerrera 
https://dotnet.myget.org/feed/msbuild/package/nuget/Microsoft.Build.Localization

**Pull request merged**